pkg/util/allocationfilterutil/v2/lexer.go

```diff
@@ -0,0 +1,272 @@
+// help the parser convert tokens into AllocationFilterConditions.
+//
+// TODO: Should the lexer _also_ attach the FilterField to help the parser out?
+var ff1ToKCFilterField = map[string]kubecost.FilterField{
+	"cluster":        kubecost.FilterClusterID,
+	"node":           kubecost.FilterNode,
+	"namespace":      kubecost.FilterNamespace,
+	"controllerName": kubecost.FilterControllerName,
+	"controllerKind": kubecost.FilterControllerKind,
+	"container":      kubecost.FilterContainer,
+	"pod":            kubecost.FilterPod,
+	"services":       kubecost.FilterServices,
+}
+var ff2ToKCFilterField = map[string]kubecost.FilterField{
+	"label":      kubecost.FilterLabel,
+	"annotation": kubecost.FilterAnnotation,
+}
+
+func (tk tokenKind) String() string {
+	switch tk {
+	case colon:
+		return "colon"
+	case comma:
+		return "comma"
+	case plus:
+		return "plus"
+	case bangColon:
+		return "bangColon"
+	case str:
+		return "str"
+	case filterField1:
+		return "filterField1"
+	case filterField2:
+		return "filterField2"
+	case keyedAccess:
+		return "keyedAccess"
+	case identifier:
+		return "identifier"
+	case eof:
+		return "eof"
+	default:
+		return fmt.Sprintf("Unspecified: %d", tk)
+	}
+}
+
+// ============================================================================
+// Lexer/Scanner
+//
+// Based on the Scanner class in Chapter 4: Scanning of Crafting Interpreters by
+// Robert Nystrom
+// ============================================================================
+
+type token struct {
+	kind tokenKind
+	s    string
+}
+
+func (t token) String() string {
+	return fmt.Sprintf("%s:%s", t.kind, t.s)
+}
+
+type scanner struct {
+	source string
+	tokens []token
+	errors []error
+
+	lexemeStartByte int
+	nextByte        int
+}
+
+func (s *scanner) scanTokens() {
+	for !s.atEnd() {
+		s.lexemeStartByte = s.nextByte
+		s.scanToken()
+	}
+
+	s.tokens = append(s.tokens, token{kind: eof})
+}
+
+func (s scanner) atEnd() bool {
+	return s.nextByte >= len(s.source)
+}
+
+// advance returns a byte because we only accept ASCII, which has to fit in a
+// byte
+func (s *scanner) advance() byte {
+	b := s.source[s.nextByte]
+	s.nextByte += 1
+	return b
+}
+
+func (s *scanner) match(expected byte) bool {
+	if s.atEnd() {
+		return false
+	}
+	if s.source[s.nextByte] != expected {
+		return false
+	}
+	s.nextByte += 1
+	return true
+}
+
+func (s *scanner) addToken(kind tokenKind) {
+	lexemeString := s.source[s.lexemeStartByte:s.nextByte]
+	switch kind {
+	// Eliminate surrounding characters like " and []
+	case str, keyedAccess:
+		lexemeString = lexemeString[1 : len(lexemeString)-1]
+	}
+
+	s.tokens = append(s.tokens, token{
+		kind: kind,
+		s:    lexemeString,
+	})
+}
+
+func (s *scanner) peek() byte {
+	if s.atEnd() {
+		return 0
+	}
+	return s.source[s.nextByte]
+}
+
+func (s *scanner) scanToken() {
+	// TODO: DON'T ACCEPT NON-ASCII
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
```

Comment:
I'm almost certain that we only need to support ASCII characters in our filters because of K8s naming conventions, but it might be possible that Prometheus or cloud provider data will be encoded in UTF-8.

If we need to support UTF-8, I'll update the lexer appropriately. If we don't, I think I should add some guard that causes the lexer to fail if a UTF-8 character is found.

Reply:
How difficult would it be to support UTF-8? Probably okay for > 99% of use cases with just ASCII.